alpha/veneer/semver/semver_test.go

```diff
@@ -0,0 +1,548 @@
+	channelOperatorVersions := semverRenderedChannelVersions{
+		"stable": {
+			"a-v0.1.0": semver.MustParse("0.1.0"),
+			"a-v0.1.1": semver.MustParse("0.1.1"),
+			"a-v1.1.0": semver.MustParse("1.1.0"),
+			"a-v1.2.1": semver.MustParse("1.2.1"),
+			"a-v1.3.1": semver.MustParse("1.3.1"),
+			"a-v2.1.0": semver.MustParse("2.1.0"),
+			"a-v2.1.1": semver.MustParse("2.1.1"),
+			"a-v2.3.1": semver.MustParse("2.3.1"),
+			"a-v2.3.2": semver.MustParse("2.3.2"),
+		},
+	}
+	// map[string]string
+	channelNameToKind := map[string]string{
+		"stable-v0":   "stable",
+		"stable-v1":   "stable",
+		"stable-v2":   "stable",
+		"stable-v0.1": "stable",
+		"stable-v1.1": "stable",
+		"stable-v1.2": "stable",
+		"stable-v1.3": "stable",
+		"stable-v2.1": "stable",
+		"stable-v2.3": "stable",
+	}
+
+	majorGeneratedUnlinkedChannels := map[string]*declcfg.Channel{
+		"stable-v0": {
+			Schema:  "olm.channel",
+			Name:    "stable-v0",
+			Package: "a",
+			Entries: []declcfg.ChannelEntry{
+				{Name: "a-v0.1.0"},
+				{Name: "a-v0.1.1"},
+			},
+		},
+		"stable-v1": {
+			Schema:  "olm.channel",
+			Name:    "stable-v1",
+			Package: "a",
+			Entries: []declcfg.ChannelEntry{
+				{Name: "a-v1.1.0"},
+				{Name: "a-v1.2.1"},
+				{Name: "a-v1.3.1"},
+			},
+		},
+		"stable-v2": {
+			Schema:  "olm.channel",
+			Name:    "stable-v2",
+			Package: "a",
+			Entries: []declcfg.ChannelEntry{
+				{Name: "a-v2.1.0"},
+				{Name: "a-v2.1.1"},
+				{Name: "a-v2.3.1"},
+				{Name: "a-v2.3.2"},
+			},
+		},
+	}
+
+	minorGeneratedUnlinkedChannels := map[string]*declcfg.Channel{
+		"stable-v0.1": {
+			Schema:  "olm.channel",
+			Name:    "stable-v0.1",
+			Package: "a",
+			Entries: []declcfg.ChannelEntry{
+				{Name: "a-v0.1.0"},
+				{Name: "a-v0.1.1"},
+			},
+		},
+		"stable-v1.1": {
+			Schema:  "olm.channel",
+			Name:    "stable-v1.1",
+			Package: "a",
+			Entries: []declcfg.ChannelEntry{
+				{Name: "a-v1.1.0"},
+			},
+		},
+		"stable-v1.2": {
+			Schema:  "olm.channel",
+			Name:    "stable-v1.2",
+			Package: "a",
+			Entries: []declcfg.ChannelEntry{
+				{Name: "a-v1.2.1"},
+			},
+		},
+		"stable-v1.3": {
+			Schema:  "olm.channel",
+			Name:    "stable-v1.3",
+			Package: "a",
+			Entries: []declcfg.ChannelEntry{
+				{Name: "a-v1.3.1"},
+			},
+		},
+		"stable-v2.1": {
+			Schema:  "olm.channel",
+			Name:    "stable-v2.1",
+			Package: "a",
+			Entries: []declcfg.ChannelEntry{
+				{Name: "a-v2.1.0"},
+				{Name: "a-v2.1.1"},
+			},
+		},
+		"stable-v2.3": {
+			Schema:  "olm.channel",
+			Name:    "stable-v2.3",
+			Package: "a",
+			Entries: []declcfg.ChannelEntry{
+				{Name: "a-v2.3.1"},
+				{Name: "a-v2.3.2"},
+			},
+		},
+		"stable-v3.1": {
+			Schema:  "olm.channel",
+			Name:    "stable-v3.1",
+			Package: "a",
+			Entries: []declcfg.ChannelEntry{
+				{Name: "a-v3.1.0"},
+				{Name: "a-v3.1.1"},
+			},
+		},
+	}
+
+	tests := []struct {
```

Comment:
It'd be nice to add a testcase for [prerelease versions](https://semver.org/#spec-item-9) and [build metadata tags](https://semver.org/#spec-item-10).

Reply:
Pre-release versions are interesting from the perspective of this veneer. I wonder if we should have a check that ensures pre-release versions are only present in `candidate*` channels?

Reply:
Introduced prerelease versioning to a couple of the utests since the sequencing is so very important, but I really have to remind folks that the original AC here did not include docs/e2e/utest.

Reply:
>Build metadata tags are intentionally not sortable according to the semver spec. What are you thinking for a test involving build metadata versions?

@joelanford This isn't entirely true, `v0.0.1` and `v0.0.1+metadata` are still sortable but they are given the same precedence.  One of the goals of the proposed schema was to be deterministic, I was hoping that to test that the order was consistent even when multiple entries with different semver build metadata were included in a channel, provided that the semver veneer schema hadn't changed. My hope that the output is consistent provided that no changes where made to the semver schema (the order of entries in the channel did not change).

Reply:
>Introduced prerelease versioning to a couple of the utests since the sequencing is so very important, but I really have to remind folks that the original AC here did not include docs/e2e/utest.

In general, we push for a lot of unit tests before passing code to other teams. However, this is an alpha feature so I'm not opposed to merging this with the set of tests you've provided, assuming you can answer the consistency question proposed in the previous comment.